### PR TITLE
Trie

### PR DESCRIPTION
**Pull request by happy-san**
#19 @code-shoily

### CODE REVIEW

**Comment by code-shoily** [comment]
What about tests?

File: lib/trie/trie.dart [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    return node;
+  }
+
+  /// Returns index, which represents the [component] in `node.children`.
+  int _indexOf(V component) => _components.indexOf(component);
+
+  List _split(V value) {
+    List<V> list = splitter(value);
+    for (var component in list) {
+      if (!components.contains(component)) throw ('$component ∉ $components');
```
Comment (code-shoily) [inline lib/trie/trie.dart:140]: Try throwing an error object instead, easily shareable and testable.

File: lib/trees/binary_tree_adt.dart [hunk 2]
```diff
@@ -38,21 +38,21 @@ abstract class BinaryTreeADT<N extends BinaryNodeADT, V extends Comparable>
...
 
-  @override
+  /// PostOrder Traversal.
   List<V> postOrder() {
     var result = <V>[];
     _postOrder(root, result);
     return result;
   }
 
-  @override
```
Comment (code-shoily) [inline lib/trees/binary_tree_adt.dart:55]: Good call 👍

**Comment by happy-san** [comment]
Yeah, will implement tests in a bit. Wasn't sure about the splitter approach. Thanks for reviewing.

File: test/trie/trie_test.dart [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+  test('Components', () {
+    expect(trie.components,
+        {...List.generate(26, (i) => String.fromCharCode(122 - i))});
+
+    expect(customTrie.components, {'f', 'o', 'b', 'a', 'r', 'z'});
+  });
+
+  test('Empty trie', () {
+    expect(emptyTrie.isEmpty, true);
```
Comment (code-shoily) [inline test/trie/trie_test.dart]: Use isTrue and isFalse instead of true and false?

File: test/trie/trie_test.dart [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    customTrie = Trie(
+        {'a', 'b', 'f', 'o', 'r', 'z'}, (value) => value.toString().split(''));
+    customTrie.add('foo');
+  });
+
+  test('Components', () {
+    expect(trie.components,
+        {...List.generate(26, (i) => String.fromCharCode(122 - i))});
+
+    expect(customTrie.components, {'f', 'o', 'b', 'a', 'r', 'z'});
```
Comment (code-shoily) [inline test/trie/trie_test.dart]: `expect(customTrie.components, equals({'f', 'o', 'b', 'a', 'r', 'z'}))`

**Review by code-shoily — APPROVED** [review]

**Pull request merged**